src/transforms/serializer.js

```diff
@@ -1,185 +1,60 @@
-var [readVarInt, writeVarInt, sizeOfVarInt] = require("../datatypes/utils").varint;
-var protocol = require("../protocol");
-var Transform = require("readable-stream").Transform;
-var debug = require("../debug");
-var assert = require('assert');
-var { getFieldInfo, tryCatch, addErrorField } = require('../utils');
+var ProtoDef = require("protodef").ProtoDef;
+var Serializer = require("protodef").Serializer;
+var Parser = require("protodef").Parser;
 
-module.exports.createSerializer = function(obj) {
-  return new Serializer(obj);
-};
-
-module.exports.createDeserializer = function(obj) {
-  return new Deserializer(obj);
-};
-
-// This is really just for the client.
-var states = {
-  "HANDSHAKING": "handshaking",
-  "STATUS": "status",
-  "LOGIN": "login",
-  "PLAY": "play"
-};
-module.exports.states = states;
-
-var NMProtocols = require("../protocol");
-var numeric = require("../datatypes/numeric");
-var utils = require("../datatypes/utils");
 var minecraft = require("../datatypes/minecraft");
-var structures = require("../datatypes/structures");
-var conditional = require("../datatypes/conditional");
-var readPackets = require("../packets").readPackets;
-
+var states = require("../states");
 
-function createProtocol(types)
+function createProtocol(types,packets)
 {
-  var proto = new NMProtocols();
-  proto.addTypes(numeric);
-  proto.addTypes(utils);
+  var proto = new ProtoDef();
   proto.addTypes(minecraft);
-  proto.addTypes(structures);
-  proto.addTypes(conditional);
   proto.addTypes(types);
+
+  Object.keys(packets).forEach(function(name) {
+    proto.addType("packet_"+name,["container",packets[name].fields]);
+  });
+
+  proto.addType("packet",["container", [
+    { "name": "id", "type": "varint" },
+    { "name": "params", "type": ["switch", {
```

Comment:
Or maybe not : we might conflict with ID. Urgh. I think it might be a good idea to rethink switch a tiny bit so it can "contain" what it switches on (id here) and avoid conflict, without making the API suck.